src/utils.rkt

```diff
@@ -1,16 +1,169 @@
+(define (bv-cost int-val)
+  (define bv-val (bitvectorize-concrete (cost-fin) int-val))
+  (assert (<= 0 (bitvector->integer bv-val))
+          (~a "Invalid integer for creating bitvector cost: " int-val))
+  bv-val)
+
+;;========================= BITVECTOR LISTS =========================
+
+(define (make-symbolic-bv-list ty size)
+  (for/list ([_ (in-range size)])
+    (define-symbolic* v ty)
+    (box v)))
+
+(define (make-bv-list-empty size)
+  (for/list ([_ (in-range size)])
+    (box void)))
+
+(define (make-symbolic-bv-list-values size)
+  (make-symbolic-bv-list (bitvector (value-fin)) size))
+
+(define (make-symbolic-bv-list-indices size)
+  (make-symbolic-bv-list (bitvector (index-fin)) size))
+
+(define (make-symbolic-matrix rows cols)
+  (matrix rows cols (make-symbolic-bv-list-values (* rows cols))))
+
+(define (make-bv-list-zeros size)
+  (for/list ([_ (in-range size)])
+    (box (bv-value 0))))
+
+(define (bv-list to-bv xs)
+  (define elements (map to-bv xs))
+  (map box elements))
+
+(define (value-bv-list . xs)
+  (bv-list bv-value xs))
+
+(define (index-bv-list . xs)
+  (bv-list bv-index xs))
+
+(define (bv-list-set! lst idx val)
+  (assert (list? lst) (~a "Expected a list, got " lst))
+  (match lst
+    [(cons box tail)
+      (if (bveq idx (bv-index 0))
+          (set-box! box val)
+          (bv-list-set! tail (bvsub idx (bv-index 1)) val))]
+    [_ (error "List idx not found" idx lst)]))
+
+(define (bv-list-get lst idx)
+  (assert (list? lst) (~a "Expected a list, got " lst))
+  (match lst
+    [(cons box tail)
+      (if (bveq idx (bv-index 0))
+          (unbox box)
+          (bv-list-get tail (bvsub idx (bv-index 1))))]
+    [_ (error "List idx not found" idx lst)]))
+
+(define (concretize-bv-list lst)
+  (define (bv-unbox x) (bitvector->integer (unbox x)))
+  (list->vector (map bv-unbox lst)))
+
+; Mutates the destination list in place, setting box values to the values boxed
+; in the given source elements
+(define (bv-list-copy! dest
+                       dest-start
+                       src
+                       [src-start (bv-index 0)]
+                       [src-end (bv-index (length src))])
+
+  (assert (list? dest) (~a "Expected a list, got " dest))
+  (assert (list? src) (~a "Expected a list, got " src))
+  (assert (bv-index? dest-start) (~a "Expected index vector, got" dest-start))
+  (assert (bv-index? src-start) (~a "Expected index vector, got" src-start))
+  (assert (bv-index? src-end) (~a "Expected index vector, got" src-end))
+  (define copy-len (bvsub src-end src-start))
+  (assert (bvslt (bv-index 0) copy-len) (~a "Copy length must be positive, got " copy-len))
+  (assert (bvsle (bvadd dest-start copy-len) (bv-index (length dest))) (~a "Copy length overflows destination, " copy-len))
+  (assert (bvsle src-end (bv-index (length src))) (~a "Copy end index overflows source " copy-len))
+
+  ; TODO(alexa): we might be able to just replace this with a for-loop if the
+  ; bounds are never symbolic
```

Comment:
Bounds should probably not be symbolic. If they are, Rosette will generate a lot of symbolic unions.